hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsSelectorCommon.java

```diff
@@ -316,6 +343,197 @@ private static Dataset<Row> coalesceOrRepartition(Dataset dataset, int numPartit
     return dataset;
   }
 
+  private static boolean isCoalesceRequired(TypedProperties properties, Schema sourceSchema) {
+    return getBooleanWithAltKeys(properties, CloudSourceConfig.SPARK_DATASOURCE_READER_COALESCE_ALIAS_COLUMNS)
+        && Objects.nonNull(sourceSchema)
+        && hasFieldWithAliases(sourceSchema);
+  }
+
+  /**
+   * Recursively checks if an Avro schema or any of its nested fields contain aliases.
+   *
+   * @param schema The Avro schema to check.
+   * @return True if the schema or any of its fields contain aliases, false otherwise.
+   */
+  private static boolean hasFieldWithAliases(Schema schema) {
+    // If the schema is a record, check its fields recursively
+    if (isNestedRecord(schema)) {
+      for (Schema.Field field : getRecordFields(schema)) {
+        // Check if the field has aliases
+        if (!field.aliases().isEmpty()) {
+          return true;
+        }
+        // Recursively check the field's schema for aliases
+        if (hasFieldWithAliases(field.schema())) {
```

Comment:
With this method, along with `getNestedFields`, and other recursive methods, there is potential for performance bottlenecks with very large and deeply nested schemas. Consider testing these methods on complex, deeply nested schemas to ensure performance remains acceptable. If you can do a performance profile that would be great. Especially, if any of these methods are being called per row then it's going to be a bottleneck.

Reply:
This code isn't called on every row. This is called once before reading data from source. I can do a performance profile and add the results here if required.